src/testRunner/unittests/tsserver/updateFileSystem.ts

```diff
@@ -0,0 +1,125 @@
+namespace ts.projectSystem {
+    describe("unittests:: tsserver:: updateFileSystem", () => {
+        const config: protocol.FileSystemRequestArgs = {
+            file: "/fshost/b/tsconfig.json",
+            fileContent: "{}"
+        };
+        const file3: protocol.FileSystemRequestArgs = {
+            file: "/fshost/b/file3.ts",
+            fileContent: "export let xyz = 1;"
+        };
+        const app: protocol.FileSystemRequestArgs = {
+            file: "/fshost/b/app.ts",
+            fileContent: "import { xyz } from './file3'; let x = xyz"
+        };
+        const file1: protocol.FileSystemRequestArgs = {
+            file: "/fshost/b/commonFile1.ts",
+            fileContent: "let x = 1"
+        };
+        const file2: protocol.FileSystemRequestArgs = {
+            file: "/fshost/b/commonFile2.ts",
+            fileContent: "let y = 1"
+        };
+        const lib: protocol.FileSystemRequestArgs = {
+        file: "/fshost/lib/lib.d.ts",
+        fileContent: `/// <reference no-default-lib="true"/>
+interface Boolean {}
+interface Function {}
+interface CallableFunction {}
+interface NewableFunction {}
+interface IArguments {}
+interface Number { toExponential: any; }
+interface Object {}
+interface RegExp {}
+interface String { charAt: any; }
+interface Array<T> { length: number; [n: number]: T; }`
+    };
+
+        function fileContentWithComment(file: protocol.FileSystemRequestArgs) {
+            return `// some copy right notice
+${file.fileContent}`;
+        }
+        function baselineFileSystem(scenario: string, subScenario: string, requests: [string, Partial<protocol.Request>][], host: VirtualFS.VirtualServerHost, session: TestSession) {
+            const history: string[] = [];
+            let prev = host.snap();
+            for (const [name, request] of requests) {
+                session.executeCommandSeq(request);
+                history.push("");
+                history.push("#### " + name);
+                host.diff(history, prev);

```

Comment:
session.host.diff
session.fsHost.dff
also serialize watches on both so we know they are on correct host?

Reply:
Timeouts on session.host so we know timeouts are set correctly. 

You need to modify unopened file in the project to show that it queues timeout for program update

Reply:
Do I need to serialise watches after every step or just at the end? I don't know how much granularity I need to capture.

Reply:
I added a method to make Callbacks show its active timeouts, but I don't know what the output is supposed to look like. Can you take a look? Currently it only shows timeouts when modifying a file on fshost, but not host.